freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/VirtualGroupSyncService.java

```diff
@@ -0,0 +1,52 @@
+package com.sequenceiq.freeipa.service.freeipa.user;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.auth.altus.UmsVirtualGroupRight;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupService;
+import com.sequenceiq.freeipa.configuration.UsersyncConfig;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UserSyncOptions;
+
+@Service
+public class VirtualGroupSyncService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(VirtualGroupSyncService.class);
+
+    @Inject
+    @Qualifier(UsersyncConfig.USERSYNC_TASK_EXECUTOR)
+    private ExecutorService asyncTaskExecutor;
+
+    @Inject
+    private VirtualGroupService virtualGroupService;
+
+    public void syncVirtualGroups(String accountId, Set<String> environmentCrns, UserSyncOptions options) {
+        if (options.isFullSync()) {
+            LOGGER.info("Sync virtual groups for environments: {}", environmentCrns);
+            Map<String, Future<Map<UmsVirtualGroupRight, String>>> userGroupFutures = environmentCrns.stream().collect(Collectors.toMap(envCrn -> envCrn, envCrn -> asyncCreateUserGroups(accountId, envCrn)));
+            userGroupFutures.forEach((envCrn, userGroupFuture) -> {
+                try {
+                    Map<UmsVirtualGroupRight, String> virtualGroups = userGroupFuture.get();
+                    LOGGER.info("Virtual group sync finished for env '{}'. Synchronized virtual groups: '{}'", envCrn, virtualGroups.values());
+                } catch (Exception e) {
+                    LOGGER.error("Virtual group sync failed for env '{}': '{}'", envCrn, e.getMessage());
+                }
+            });
+        }
+    }
+
+    private Future<Map<UmsVirtualGroupRight, String>> asyncCreateUserGroups(String accountId, String environmentCrn) {
+        return asyncTaskExecutor.submit(() -> virtualGroupService.createVirtualGroups(accountId, environmentCrn));
```

Comment:
@lacikaaa Do you think we should fire async calls or we can simply initiate sync calls? This method is triggered from UserSyncService...

Reply:
how long does it take to create those groups?

Reply:
I am curious about it, let me check

Reply:
we are planning to create another executor to separate the 2 kind of async tasks.
- the overseeing operation
- the actual sync calls

this is the latter, but maybe if these calls are very quick and low in number, like we need 10 of these calls at max and takes 1 sec per request then it's fine without another async call. if every call takes like 30 sec and we need a 100 of them, then we need to make those async

Reply:
![Screenshot from 2022-03-31 09-19-12](https://user-images.githubusercontent.com/26329499/160999805-613de6cb-555e-436e-bf32-b8de39b98d9e.png)

considering the worst possible case we would make (1 get virtual group call + 1 set virtual group call) * virtual group count = 18 similar UMS call (assuming that @pkedvessy or his team will introduce one new right for the sudoer feature)

so calling these sequentially would take: 5 ms (avg get call) * 9 + 35 ms (avg set call) * 9 = 360 ms

Reply:
okey, seems like we don't need async calls as the performance impact is minimal

Reply:
to be precise, I'll add stats about dev:
![Screenshot from 2022-03-31 09-32-14](https://user-images.githubusercontent.com/26329499/161001992-7ce6cb97-aac6-49d0-a117-e9e25af37b14.png)

anyway the real use cases here after deployment if this feature:
- for new environments there will be 9 get calls during sync, set calls will made during environment creation flow based on this implementation
- for existing environments, there will be 9 get calls and 1 set call for new right during first sync after deployment of this new feature, the case is similar, one longer set call only

Reply:
@pkedvessy it also seems that during creation we would not need an extra get call:
from proto file:
```
// Sets the workload administration group name for the (account, right, resource)
  // tuple. If the name was already set for the workload administration group
  // this is a no-op and the name generated for the workload administration group
  // will be returned.
  rpc SetWorkloadAdministrationGroupName(SetWorkloadAdministrationGroupNameRequest)
    returns (SetWorkloadAdministrationGroupNameResponse) {}
```
might worth an optimization

Reply:
not sure though, since get calls are faster, so if group exists, the method would finish sooner

Reply:
I think it would worth adding `Benchmark#measure` when calling this, and we would have numbers about how long does this take in mow-* environments. But overall I agree that async seems to be an overkill here

Reply:
done, thanks again for checking it